### ISSUE TITLE
When a rule is enabled in a child configuration, but configured in the config referencing that child, swift reports that the rule is not enabled

### ISSUE THREAD

**Issue by mildm8nnered** [realm/SwiftLint#5251]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Originally reported in #4858 by @tinder-garricnahapetian - See https://github.com/realm/SwiftLint/issues/4858#issuecomment-1721651190

When a rule is enabled in a child configuration, but configured in the config that references that child, SwiftLint will report:

`warning: Found a configuration for 'empty_count' rule, but it is not enabled on 'opt_in_rules'.`

even though the rule will actually be enabled in the eventual configuration.

This is related to, but distinct from #4858.

In #4858, we were simply printing incorrect disablement messages, when rules were actually enabled. This logic was relatively easy to fix to correct this (see #4864).

But when `child_config`'s are used, the situation is slightly different.

The configurations are initialized individually, which is when the warnings are printed. Then the configurations are merged.

So to fix this, we would either need to change how configurations are merged (for example, resolving `child_config` references at initialization), or when we do validation of configurations and enabled rules - moving it to after all configurations have been resolved and merged.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
warning: Found a configuration for 'empty_count' rule, but it is not enabled on 'opt_in_rules'.
Linting Swift files in current working directory
Error: No lintable files found at paths: ''
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Homebrew
* Paste your configuration file:

I'm seeing this as well. Main difference is that we are using only_rules, not opt_in_rules.

Found a configuration for 'empty_count' rule, but it is not enabled on 'opt_in_rules'.

Configs:

```
% cat .swiftlint.yml

included:
- Sources

child_config: .swiftlint-rules.yml

empty_count:
  only_after_dot: true
```

```
 % cat .swiftlint-rules.yml
only_rules:
- empty_count
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)? 

```
Xcode 15.0
Build version 15A240d
```

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

See above

**Comment by tinder-garricnahapetian** [comment]
Thank you for opening this issue.
